[nativert] port enumerate from folly to c10::utill

dolpm opened the pull request:
Summary:
nativert RFC: https://github.com/zhxchen17/rfcs/blob/master/RFC-0043-torch-native-runtime.md

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff ports an enumeration util from folly into c10.

Test Plan: CI

Differential Revision: D73881042

cyyever commented:
I would rather add ranges-v3 as a submodule and be able to use all range functions...

zhxchen17 commented:
Need to fix lint, otherwise looks good to me.

zhxchen17 commented:
> I would rather add ranges-v3 as a submodule and be able to use all range functions...

IMO it'd be better if we can enable c++20/23 directly cuz nativert is using features more than ranges.

cyyever commented:
> > I would rather add ranges-v3 as a submodule and be able to use all range functions...
> 
> IMO it'd be better if we can enable c++20/23 directly cuz nativert is using features more than ranges.

It's not likely to happen this year, may be next year..

dolpm committed:
[nativert] port enumerate from folly to c10::util (#152481)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/152481

nativert RFC: https://github.com/zhxchen17/rfcs/blob/master/RFC-0043-torch-native-runtime.md

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff ports an enumeration util from folly into c10.

Test Plan: CI

Reviewed By: zhxchen17

Differential Revision: D73881042
